db/counting_index.go

```diff
@@ -41,41 +42,52 @@ type (
 		Commit() error
 	}
 
-	// countingIndex is CountingIndex implementation based on bolt DB
+	// countingIndex is CountingIndex implementation based on KVStore
 	countingIndex struct {
-		db         *bolt.DB
-		numRetries uint8
-		bucket     []byte
-		size       uint64 // total number of keys
-		batch      KVStoreBatch
-	}
-
-	// memCountingIndex is the in-memory implementation of CountingIndex
-	memCountingIndex struct {
-		kvstore *memKVStore
-		bucket  []byte
+		kvStore KVStore
+		bucket  string
 		size    uint64 // total number of keys
 		batch   KVStoreBatch
 	}
 )
 
-// NewCountingIndex creates a new instance of countingIndex
-func NewCountingIndex(db *bolt.DB, retry uint8, name []byte, size uint64) (CountingIndex, error) {
-	if db == nil {
-		return nil, errors.Wrap(ErrInvalid, "db object is nil")
+// CreateCountingIndexNX creates a new counting index if it does not exist, otherwise return existing index
+func CreateCountingIndexNX(kv KVStore, name []byte) (CountingIndex, error) {
```

Comment:
define name as "string"?

Reply:
would keep the same
1. name is account address which is []byte
2. down the callstack, the string is actually converted to []byte to form the key before writing to DB